components/omega/doc/design/DataTypes.md

```diff
@@ -0,0 +1,108 @@
+
+# OMEGA Requirements and Design:
+
+# *DataTypes*
+
+
+## 1 Overview
+
+Data type aliases are a useful means for managing precision within OMEGA and
+to improve readability for for YAKL array data types. This header file 
+defines a number of aliases to common data types used throughout OMEGA.
+
+
+## 2 Requirements
+
+
+### 2.1 Requirement: Both flexible and fixed floating point
+
+We wish to explore mixed precision and lower precision floating point within
+OMEGA so require both a generic floating point type that can be switched at 
+compile time as well as explicit single and double-precision (32-bit, 64-bit) 
+types to enforce precision where needed.  
+
+### 2.2 Desired: Readability for array types
+
+OMEGA will be using YAKL array data types which have a long syntax. For
+readability, these should be aliased to a shorter, intuitive data type.
+
+### 2.3 Desired: Fixed width integer types
+
+For better control over integer storage, it might be better to define
+fixed-width integer data types rather than the generic int, long.
+
+### 2.4 Future: Floating point conversion
+
+Currently, language standard floating point conversion should be adequate,
+but future mixed-precision implementations may require more reproducible
+methods for implementation-dependent conversion. 
+
+## 3 Algorithmic Formulation
+
+No algorithms needed yet, though reproducibility in type conversion may
+require reproducible algorithms in the future.
+
+## 4 Design
+
+Currently, the data types can be defined in a single header file that sets
+up aliases.
+
+### 4.1 Data types and parameters
+
+#### 4.1.1 Parameters 
+
+We will use a global cpp define -D SINGLE_PRECISION at compile time as a 
+global switch for using single precision as the default real data type. 
+Otherwise, the default real will be double precision. 
+
+
+#### 4.1.2 Class/structs/data types
+
+The data types will be defined using type aliases within a single header 
+file DataTypes.h We will use the "using" syntax rather than 
+the older typedef.
+
+
+    // Standard integer and floating point types
+    using I4 = std::int32_t;
+    using I8 = std::int64_t;
+    using R4 = float;
+    using R8 = double;
+    #ifdef SINGLE_PRECISION
+    using Real = float;
+    #else
+    using Real = double;
+    #endif
+
+    // Aliases for YAKL arrays.
+    using Array1DI4   = YAKL::Array<I4,1,memDevice,styleC>
+    using Array1DI8   = YAKL::Array<I8,1,memDevice,styleC>
+    using Array1DR4   = YAKL::Array<R4,1,memDevice,styleC>
+    using Array1DR8   = YAKL::Array<R8,1,memDevice,styleC>
+    using Array1DReal = YAKL::Array<Real,1,memDevice,styleC>
+    using Array2DI4   = YAKL::Array<I4,2,memDevice,styleC>
+    using Array2DI8   = YAKL::Array<I8,2,memDevice,styleC>
+    using Array2DR4   = YAKL::Array<R4,2,memDevice,styleC>
+    using Array2DR8   = YAKL::Array<R8,2,memDevice,styleC>
+    using Array2DReal = YAKL::Array<Real,2,memDevice,styleC>
+    // continue this pattern for higher-dimensional arrays
+
```

Comment:
This came up during the meeting with Matt, but we will probably want to make a decision on C-style vs Fortran-style arrays sooner rather than later. I thought one of the reasons we decided upon YAKL was the Fortran-style array option, and to echo Matt, if we go with C-style that introduces a potential pitfall to worry about at every step when we get to the actual porting process.

Though if we are writing from scratch in C++, and given broader trends, there is an argument to make a clean break and embrace the C++ lifestyle. Also, I think we're all primarily Fortran coders, but users or future developers who are more C-focused and not familiar with YAKL may be confused by Fortran-style arrays in a C++ code.

Reply:
Thanks Brian and yes, we probably need to take this up at the next Omega meeting.  I agree that it will make the short term porting easier and reduce the chance of introducing indexing errors.  But if we decide we want to switch to C-style later, it will mean another pretty big code effort and the same risk of error later. 

BTW, also as a result of that meeting, I need to add aliases for Host arrays too. I'll add that once we decide which way we want to go.

Reply:
FWIW, I vote for C style arrays as we lay the groundwork now. To paraphrase Pat's writing Fortran in any language, we should write "C in C++" to an extent where we selective utilize the language features for performance and readability reasons.  

It would minimize any of the eccentricities of the codebase in the future. Additionally, we will be more thorough with testing (V&V) at this stage.

Reply:
No objections on Omega call to C-ordering so will leave as is.